File: cpp/tests/test_propulsion.cpp
// Author: A Taylor | Purpose: Unit tests for gridded ion thruster model | Ref: Goebel/Katz/Vallado

#include <gtest/gtest.h>
#include "ion_propulsion/propulsion/thruster.hpp"
#include "ion_propulsion/constants.hpp"

#include <cmath>
#include <stdexcept>

using namespace ion_propulsion::propulsion;
using namespace ion_propulsion::constants;

namespace {

constexpr double rel_tol = 1e-6;

/// Cross-language golden values (identical in the Python and MATLAB suites).
constexpr double golden_exhaust_velocity_ms      = 40210.39047562004;
constexpr double golden_mass_utilization         = 0.7982474005630351;
constexpr double golden_electrical_efficiency    = 0.8461538461538461;
constexpr double golden_extraction_efficiency    = 0.675440108168722;
constexpr double golden_specific_impulse_s       = 3273.068751590841;
constexpr double golden_thrust_N                 = 0.09629351901836497;
constexpr double golden_beam_power_W             = 1936.0;
constexpr double golden_total_power_W            = 2288.0;
constexpr double golden_child_langmuir_Am2       = 218.48574875215593;
constexpr double golden_grid_erosion_kgs         = 1.7510179217855205e-07;
constexpr double golden_neutralizer_W            = 40.0;

/// NSTAR flight data at full power (Deep Space 1).
constexpr double nstar_flight_isp_s    = 3100.0;
constexpr double nstar_flight_thrust_N = 0.092;
constexpr double nstar_flight_power_W  = 2300.0;
constexpr double benchmark_tol         = 0.10;  // divergence and double ions not modelled

constexpr double nstar_mass_flow_kgs = 3.0e-6;

/// NSTAR at full power: 1100 V beam, 1.76 A, 3.0 mg/s xenon, accel grid at -180 V.
GriddedIonThruster make_nstar() {
    return GriddedIonThruster(1100.0, 1.76, 1100.0, -180.0, nstar_mass_flow_kgs);
}

/// 1500 V operating point used for the Isp >= 3000 s design requirement.
GriddedIonThruster make_high_voltage() {
    return GriddedIonThruster(1500.0, 1.8, 1500.0, 250.0, 3.5e-6, m_xenon, 0.7e-3, 0.95e-3);
}

#define EXPECT_REL_NEAR(actual, expected) \
    EXPECT_NEAR((actual), (expected), rel_tol * std::fabs(expected))

} // anonymous namespace

// ---------------------------------------------------------------------------
// NSTAR benchmark
// ---------------------------------------------------------------------------
TEST(PropulsionTest, Nstar_IspWithin10PercentOfFlight) {
    const double isp = make_nstar().specific_impulse();
    EXPECT_LT(std::fabs(isp - nstar_flight_isp_s) / nstar_flight_isp_s, benchmark_tol);
}

TEST(PropulsionTest, Nstar_ThrustWithin10PercentOfFlight) {
    const double F = make_nstar().thrust();
    EXPECT_LT(std::fabs(F - nstar_flight_thrust_N) / nstar_flight_thrust_N, benchmark_tol);
}

TEST(PropulsionTest, Nstar_InputPowerWithin10PercentOfFlight) {
    const double P = make_nstar().total_power();
    EXPECT_LT(std::fabs(P - nstar_flight_power_W) / nstar_flight_power_W, benchmark_tol);
}

// ---------------------------------------------------------------------------
// Cross-language golden values
// ---------------------------------------------------------------------------
TEST(PropulsionTest, Golden_Efficiencies) {
    const auto t = make_nstar();
    EXPECT_REL_NEAR(t.mass_utilization(),          golden_mass_utilization);
    EXPECT_REL_NEAR(t.electrical_efficiency(),     golden_electrical_efficiency);
    EXPECT_REL_NEAR(t.ion_extraction_efficiency(), golden_extraction_efficiency);
}

TEST(PropulsionTest, Golden_ExhaustVelocityAndIsp) {
    const auto t = make_nstar();
    EXPECT_REL_NEAR(t.exhaust_velocity(), golden_exhaust_velocity_ms);
    EXPECT_REL_NEAR(t.specific_impulse(), golden_specific_impulse_s);
}

TEST(PropulsionTest, Golden_Thrust) {
    EXPECT_REL_NEAR(make_nstar().thrust(), golden_thrust_N);
}

TEST(PropulsionTest, Golden_Power) {
    const auto t = make_nstar();
    EXPECT_REL_NEAR(t.beam_power(),  golden_beam_power_W);
    EXPECT_REL_NEAR(t.total_power(), golden_total_power_W);
}

TEST(PropulsionTest, Golden_ChildLangmuir) {
    EXPECT_REL_NEAR(make_nstar().child_langmuir_current(), golden_child_langmuir_Am2);
}

TEST(PropulsionTest, Golden_ErosionAndNeutralizer) {
    const auto t = make_nstar();
    EXPECT_REL_NEAR(t.grid_erosion_rate(), golden_grid_erosion_kgs);
    EXPECT_REL_NEAR(t.neutralizer_power(), golden_neutralizer_W);
}

// ---------------------------------------------------------------------------
// Physics consistency
// ---------------------------------------------------------------------------
TEST(PropulsionTest, HighVoltage_IspAtLeast3000s) {
    EXPECT_GE(make_high_voltage().specific_impulse(), 3000.0);
}

TEST(PropulsionTest, Isp_NeverExceedsIdeal) {
    const auto t = make_high_voltage();
    EXPECT_LE(t.specific_impulse(), t.exhaust_velocity() / g0);
}

TEST(PropulsionTest, Thrust_EqualsEtaMassFlowExhaustVelocity) {
    const auto t = make_nstar();
    const double expected = t.mass_utilization() * nstar_mass_flow_kgs * t.exhaust_velocity();
    EXPECT_REL_NEAR(t.thrust(), expected);
}

TEST(PropulsionTest, Efficiencies_Bounded) {
    const auto t = make_nstar();
    EXPECT_GT(t.mass_utilization(), 0.0);
    EXPECT_LE(t.mass_utilization(), 1.0);
    EXPECT_GT(t.electrical_efficiency(), 0.0);
    EXPECT_LT(t.electrical_efficiency(), 1.0);
}

TEST(PropulsionTest, ChildLangmuir_AccelSignConventionIrrelevant) {
    const GriddedIonThruster neg(1100.0, 1.76, 1100.0, -180.0, nstar_mass_flow_kgs);
    const GriddedIonThruster pos(1100.0, 1.76, 1100.0,  180.0, nstar_mass_flow_kgs);
    EXPECT_DOUBLE_EQ(neg.child_langmuir_current(), pos.child_langmuir_current());
}

TEST(PropulsionTest, ChildLangmuir_NonPositiveTotalVoltageThrows) {
    const GriddedIonThruster t(1100.0, 1.76, -500.0, 100.0, nstar_mass_flow_kgs);
    EXPECT_THROW((void)t.child_langmuir_current(), std::invalid_argument);
}

// ---------------------------------------------------------------------------
// Input validation
// ---------------------------------------------------------------------------
TEST(PropulsionTest, Constructor_RejectsInvalidInputs) {
    EXPECT_THROW(GriddedIonThruster(0.0,    1.76, 1100.0, -180.0, 3.0e-6), std::invalid_argument);
    EXPECT_THROW(GriddedIonThruster(1100.0, 0.0,  1100.0, -180.0, 3.0e-6), std::invalid_argument);
    EXPECT_THROW(GriddedIonThruster(1100.0, 1.76, 1100.0, -180.0, 0.0),    std::invalid_argument);
    EXPECT_THROW(GriddedIonThruster(1100.0, 1.76, 1100.0, -180.0, 3.0e-6, m_xenon, 0.0), std::invalid_argument);
    EXPECT_THROW(GriddedIonThruster(1100.0, 1.76, 1100.0, -180.0, 3.0e-6, 0.0),     std::invalid_argument);
    EXPECT_THROW(GriddedIonThruster(1100.0, 1.76, 1100.0, -180.0, 3.0e-6, m_xenon, 1e-3, 1e-3, -1.0),
                 std::invalid_argument);
}

TEST(PropulsionTest, SputterYield_MustBePositive) {
    EXPECT_THROW((void)make_nstar().grid_erosion_rate(0.0), std::invalid_argument);
}

TEST(PropulsionTest, KeeperParameters_MustBePositive) {
    EXPECT_THROW((void)make_nstar().neutralizer_power(0.0, 2.0), std::invalid_argument);
}

TEST(PropulsionTest, Neutralizer_CustomKeeper) {
    EXPECT_REL_NEAR(make_nstar().neutralizer_power(15.0, 1.5), 22.5);
}

File: matlab/tests/test_propulsion.m
% Author: A Taylor | Purpose: Unit tests for propulsion module | Ref: Goebel/Katz/Vallado
classdef test_propulsion < matlab.unittest.TestCase
% TEST_PROPULSION Unit tests for the GriddedIonThruster class.
%
%   Validates the model against NSTAR flight data and against golden
%   values shared verbatim with the Python and C++ suites (1e-6 relative).

    properties (Constant)
        RelTol = 1e-6

        % NSTAR at full power: 1100 V beam, 1.76 A, 3.0 mg/s xenon, accel -180 V
        NstarFlightIsp    = 3100.0   % s
        NstarFlightThrust = 0.092    % N
        NstarFlightPower  = 2300.0   % W
        BenchmarkTol      = 0.10     % divergence and double ions not modelled
        NstarMassFlow     = 3.0e-6   % kg/s

        % Cross-language golden values
        GoldenExhaustVelocity = 40210.39047562004
        GoldenMassUtilization = 0.7982474005630351
        GoldenElectricalEff   = 0.8461538461538461
        GoldenExtractionEff   = 0.675440108168722
        GoldenIsp             = 3273.068751590841
        GoldenThrust          = 0.09629351901836497
        GoldenBeamPower       = 1936.0
        GoldenTotalPower      = 2288.0
        GoldenChildLangmuir   = 218.48574875215593
        GoldenErosion         = 1.7510179217855205e-07
        GoldenNeutralizer     = 40.0
    end

    properties
        nstar        % NSTAR full-power thruster
        highVoltage  % 1500 V operating point for the Isp >= 3000 s requirement
    end

    methods (TestMethodSetup)
        function createThrusters(testCase)
            testCase.nstar = GriddedIonThruster(1100, 1.76, 1100, -180, ...
                testCase.NstarMassFlow);
            testCase.highVoltage = GriddedIonThruster(1500, 2.0, 1100, -200, ...
                3.5e-6, 2.18e-25, 1e-3, 1e-3, 200);
        end
    end

    methods (Test)

        % ---- NSTAR benchmark --------------------------------------------
        function test_nstar_isp_within_10pct_of_flight(testCase)
            Isp = testCase.nstar.specific_impulse();
            testCase.verifyLessThan(abs(Isp - testCase.NstarFlightIsp) / testCase.NstarFlightIsp, ...
                testCase.BenchmarkTol);
        end

        function test_nstar_thrust_within_10pct_of_flight(testCase)
            F = testCase.nstar.thrust();
            testCase.verifyLessThan(abs(F - testCase.NstarFlightThrust) / testCase.NstarFlightThrust, ...
                testCase.BenchmarkTol);
        end

        function test_nstar_input_power_within_10pct_of_flight(testCase)
            P = testCase.nstar.total_power();
            testCase.verifyLessThan(abs(P - testCase.NstarFlightPower) / testCase.NstarFlightPower, ...
                testCase.BenchmarkTol);
        end

        % ---- Cross-language golden values -------------------------------
        function test_golden_efficiencies(testCase)
            t = testCase.nstar;
            testCase.verifyEqual(t.mass_utilization(),          testCase.GoldenMassUtilization, 'RelTol', testCase.RelTol);
            testCase.verifyEqual(t.electrical_efficiency(),     testCase.GoldenElectricalEff,   'RelTol', testCase.RelTol);
            testCase.verifyEqual(t.ion_extraction_efficiency(), testCase.GoldenExtractionEff,   'RelTol', testCase.RelTol);
        end

        function test_golden_exhaust_velocity_and_isp(testCase)
            t = testCase.nstar;
            testCase.verifyEqual(t.exhaust_velocity(), testCase.GoldenExhaustVelocity, 'RelTol', testCase.RelTol);
            testCase.verifyEqual(t.specific_impulse(), testCase.GoldenIsp,             'RelTol', testCase.RelTol);
        end

        function test_golden_thrust(testCase)
            testCase.verifyEqual(testCase.nstar.thrust(), testCase.GoldenThrust, 'RelTol', testCase.RelTol);
        end

        function test_golden_power(testCase)
            t = testCase.nstar;
            testCase.verifyEqual(t.beam_power(),  testCase.GoldenBeamPower,  'RelTol', testCase.RelTol);
            testCase.verifyEqual(t.total_power(), testCase.GoldenTotalPower, 'RelTol', testCase.RelTol);
        end

        function test_golden_child_langmuir(testCase)
            testCase.verifyEqual(testCase.nstar.child_langmuir_current(), ...
                testCase.GoldenChildLangmuir, 'RelTol', testCase.RelTol);
        end

        function test_golden_erosion_and_neutralizer(testCase)
            t = testCase.nstar;
            testCase.verifyEqual(t.grid_erosion_rate(), testCase.GoldenErosion,     'RelTol', testCase.RelTol);
            testCase.verifyEqual(t.neutralizer_power(), testCase.GoldenNeutralizer, 'RelTol', testCase.RelTol);
        end

        % ---- Physics consistency ----------------------------------------
        function test_high_voltage_isp_at_least_3000(testCase)
            testCase.verifyGreaterThanOrEqual(testCase.highVoltage.specific_impulse(), 3000, ...
                'Isp for a 1500 V beam must be >= 3000 s');
        end

        function test_isp_never_exceeds_ideal(testCase)
            t = testCase.highVoltage;
            testCase.verifyLessThanOrEqual(t.specific_impulse(), t.exhaust_velocity() / 9.80665);
        end

        function test_thrust_equals_eta_m_mdot_vb(testCase)
            t = testCase.nstar;
            expected = t.mass_utilization() * testCase.NstarMassFlow * t.exhaust_velocity();
            testCase.verifyEqual(t.thrust(), expected, 'RelTol', testCase.RelTol);
        end

        function test_efficiencies_bounded(testCase)
            t = testCase.nstar;
            testCase.verifyGreaterThan(t.mass_utilization(), 0);
            testCase.verifyLessThanOrEqual(t.mass_utilization(), 1);
            testCase.verifyGreaterThan(t.electrical_efficiency(), 0);
            testCase.verifyLessThan(t.electrical_efficiency(), 1);
        end

        function test_child_langmuir_sign_convention_irrelevant(testCase)
            neg = GriddedIonThruster(1100, 1.76, 1100, -180, testCase.NstarMassFlow);
            pos = GriddedIonThruster(1100, 1.76, 1100,  180, testCase.NstarMassFlow);
            testCase.verifyEqual(neg.child_langmuir_current(), pos.child_langmuir_current());
        end

        function test_child_langmuir_nonpositive_total_voltage_errors(testCase)
            t = GriddedIonThruster(1100, 1.76, -500, 100, testCase.NstarMassFlow);
            testCase.verifyError(@() t.child_langmuir_current(), ...
                'GriddedIonThruster:InvalidVoltage');
        end

        % ---- Input validation -------------------------------------------
        function test_constructor_rejects_invalid_inputs(testCase)
            testCase.verifyError(@() GriddedIonThruster(0,    1.76, 1100, -180, 3e-6), 'MATLAB:validators:mustBePositive');
            testCase.verifyError(@() GriddedIonThruster(1100, 0,    1100, -180, 3e-6), 'MATLAB:validators:mustBePositive');
            testCase.verifyError(@() GriddedIonThruster(1100, 1.76, 1100, -180, 0),    'MATLAB:validators:mustBePositive');
            testCase.verifyError(@() GriddedIonThruster(1100, 1.76, 1100, -180, 3e-6, 2.18e-25, 0), 'MATLAB:validators:mustBePositive');
            testCase.verifyError(@() GriddedIonThruster(1100, 1.76, 1100, -180, 3e-6, 0), 'MATLAB:validators:mustBePositive');
            testCase.verifyError(@() GriddedIonThruster(1100, 1.76, 1100, -180, 3e-6, 2.18e-25, 1e-3, 1e-3, -1), 'MATLAB:validators:mustBeNonnegative');
        end

        function test_sputter_yield_must_be_positive(testCase)
            t = testCase.nstar;
            testCase.verifyError(@() t.grid_erosion_rate(0), 'MATLAB:validators:mustBePositive');
        end

        function test_keeper_parameters_must_be_positive(testCase)
            t = testCase.nstar;
            testCase.verifyError(@() t.neutralizer_power(0, 2), 'MATLAB:validators:mustBePositive');
        end

        function test_neutralizer_custom_keeper(testCase)
            testCase.verifyEqual(testCase.nstar.neutralizer_power(15, 1.5), 22.5, 'RelTol', testCase.RelTol);
        end

        function test_default_arguments(testCase)
            t = GriddedIonThruster(1100, 1.76, 1100, -180, 3e-6);
            testCase.verifyEqual(t.propellant_mass, 2.18e-25, 'AbsTol', 1e-30);
            testCase.verifyEqual(t.grid_spacing, 1e-3);
            testCase.verifyEqual(t.discharge_loss, 200);
            testCase.verifyEqual(t.grid_erosion_rate(), t.grid_erosion_rate(0.1));
            testCase.verifyEqual(t.neutralizer_power(), t.neutralizer_power(20, 2));
        end

    end
end
